Projects/b_u585i_iot02a_ntz/Src/ota_pal/ota_pal_stm32u5_ntz.c

```diff
@@ -375,390 +591,570 @@ static HAL_StatusTypeDef prvWriteToFlash( uint32_t destination,
+                                               pucHashBuffer,
+                                               uxHashLength );
+        }
 
-        if( signatureValidationStatus == pdPASS )
+        if( xOtaStatus == OtaPalSuccess )
         {
-            pContext->state = OTA_PAL_IMAGE_STATE_PENDING_COMMIT;
-            otaStatus = OtaPalSuccess;
+            pxContext->xPalState = OTA_PAL_PENDING_ACTIVATION;
         }
         else
         {
-            otaStatus = OTA_PAL_COMBINE_ERR( OtaPalSignatureCheckFailed, signatureValidationStatus );
+            otaPal_SetPlatformImageState( pxFileContext, OtaImageStateRejected );
         }
     }
+    else
+    {
+    }
 
-    return otaStatus;
+    return xOtaStatus;
 }
 
 
-OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const pFileContext )
+OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const pxFileContext )
 {
-    OtaImageContext_t * pContext = prvGetImageContext();
+    return otaPal_SetPlatformImageState( pxFileContext, OtaImageStateAborted );
+}
+
+OtaPalStatus_t otaPal_ActivateNewImage( OtaFileContext_t * const pxFileContext )
+{
+    OtaPalStatus_t xOtaStatus = OtaPalActivateFailed;
 
-    if( pContext->state != OTA_PAL_IMAGE_STATE_UNKNOWN )
+    OtaPalContext_t * pxContext = prvGetImageContext();
+
+    LogInfo( "ActivateNewImage xPalState: %s", pcPalStateToString( pxContext->xPalState ) );
+
+    if( ( pxContext != NULL ) &&
+        ( pxContext->xPalState == OTA_PAL_PENDING_ACTIVATION ) )
     {
-        /*
-         * Erase the bank and set the image state to invalid, if its being programmed or
-         * pending reboot after validation.
-         */
-        if( pFileContext->pFile != NULL )
+        /** Save current image context to flash so as to fetch the context after boot. */
+        if( prvWritePalNvContext( pxContext ) == pdTRUE )
         {
-            ( void ) prvEraseBank( pContext->ulBank );
-            pFileContext->pFile = NULL;
+            if( prvSelectBank( pxContext->ulTargetBank ) == pdTRUE )
+            {
+                pxContext->xPalState = OTA_PAL_PENDING_SELF_TEST;
+                ( void ) prvWritePalNvContext( pxContext );
+                pxContext->xObLaunchPending = pdTRUE;
+                xOtaStatus = otaPal_ResetDevice( pxFileContext );
+            }
+            else
+            {
+                /* Failed to select Bank */
+                prvDeletePalNvContext();
+            }
         }
-
-        pContext->state = OTA_PAL_IMAGE_STATE_INVALID;
     }
 
-    return OtaPalSuccess;
+    return xOtaStatus;
 }
 
-OtaPalStatus_t otaPal_ActivateNewImage( OtaFileContext_t * const pFileContext )
+void otaPal_EarlyInit( void )
 {
-    OtaPalStatus_t otaStatus = OtaPalActivateFailed;
-    BaseType_t status = pdFALSE;
-    OtaImageContext_t * pContext = prvGetImageContext();
+    OtaPalContext_t * pxCtx = prvGetImageContext();
+
+    ulBankAtBootup = prvGetActiveBank();
```

Comment:
Seems the variable `ulBankAtBootup` is of global scope and is being changed inside the function `prvGetActiveBank()` as well.